src/gc/impl/proto/gc.d

```diff
@@ -0,0 +1,192 @@
+
+module gc.impl.proto.gc;
+
+import gc.config;
+import gc.gcinterface;
+
+import rt.util.container.array;
+
+import cstdlib = core.stdc.stdlib : calloc, free, malloc, realloc;
+static import core.memory;
+
+extern (C) void onOutOfMemoryError(void* pretend_sideffect = null) @trusted pure nothrow @nogc; /* dmd @@@BUG11461@@@ */
+
+private
+{
+    extern (C) void gc_init_nothrow() nothrow @nogc;
+    extern (C) void gc_term();
+
+    extern (C) void gc_enable() nothrow;
+    extern (C) void gc_disable() nothrow;
+
+    extern (C) void*    gc_malloc( size_t sz, uint ba = 0, const TypeInfo = null ) pure nothrow;
+    extern (C) void*    gc_calloc( size_t sz, uint ba = 0, const TypeInfo = null ) pure nothrow;
+    extern (C) BlkInfo gc_qalloc( size_t sz, uint ba = 0, const TypeInfo = null ) pure nothrow;
+    extern (C) void*    gc_realloc( void* p, size_t sz, uint ba = 0, const TypeInfo = null ) pure nothrow;
+    extern (C) size_t   gc_reserve( size_t sz ) nothrow;
+
+    extern (C) void gc_addRange( void* p, size_t sz, const TypeInfo ti = null ) nothrow @nogc;
+    extern (C) void gc_addRoot( void* p ) nothrow @nogc;
+}
+
+class ProtoGC : GC
+{
+    __gshared Array!Root roots;
+    __gshared Array!Range ranges;
+
+    this()
+    {
+    }
+
+    void Dtor()
+    {
+    }
+
+    void enable()
+    {
+        gc_init_nothrow();
```

Comment:
Better add an internal `initGC` method which must also replay all addRange/addRoot calls.

Reply:
@MartinNowak please not